spark3/src/main/java/org/apache/iceberg/spark/FileRewriteCoordinator.java

```diff
@@ -0,0 +1,169 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.HasTableOperations;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableOperations;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableSet;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.util.Pair;
+import org.apache.iceberg.util.Tasks;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FileRewriteCoordinator {
+
+  private static final Logger LOG = LoggerFactory.getLogger(FileRewriteCoordinator.class);
+  private static final FileRewriteCoordinator INSTANCE = new FileRewriteCoordinator();
+
+  private final Map<String, Lock> lockMap = Maps.newConcurrentMap();
+  private final Map<Pair<String, String>, Set<DataFile>> resultMap = Maps.newConcurrentMap();
+
+  private FileRewriteCoordinator() {
+  }
+
+  public static FileRewriteCoordinator get() {
+    return INSTANCE;
+  }
+
+  public void stageRewrite(Table table, String fileSetID, Set<DataFile> newDataFiles) {
+    Preconditions.checkArgument(newDataFiles != null && newDataFiles.size() > 0, "Cannot stage null or empty files");
+    LOG.info("Staging results of rewriting file set {} for table {}", fileSetID, table);
+    Pair<String, String> id = toID(table, fileSetID);
+    resultMap.put(id, newDataFiles);
+  }
+
+  public void commitRewrite(Table table, String fileSetID) {
+    commitRewrite(table, ImmutableSet.of(fileSetID));
+  }
+
+  public void commitRewrite(Table table, Set<String> fileSetIDs) {
+    LOG.info("Committing rewrite of file sets {} for table {}", fileSetIDs, table);
+
+    Set<DataFile> rewrittenDataFiles = fetchRewrittenDataFiles(table, fileSetIDs);
+    Set<DataFile> newDataFiles = fetchNewDataFiles(table, fileSetIDs);
+
+    String tableUUID = tableUUID(table);
+    Lock tableLock = lockMap.computeIfAbsent(tableUUID, uuid -> new ReentrantLock());
+
+    try {
+      LOG.info("Locking table {} to commit rewrite of file sets {}", table, fileSetIDs);
```

Comment:
Locking is important as we may be processing and committing chunks concurrently. This will avoid conflicts.

Reply:
I'll a test for this now.

Reply:
I was going to ask, I'm not sure I understand, what would the issue be with concurrent chunk commits? Shouldn't table operations block/lock?

Reply:
Let me think. We had this internally as each Spark job was doing an actual commit. It seems it still applies to the new logic if we will be doing commits from each compaction thread.

Reply:
I guess it now depends on how we are going to issue commits from the action if the user allows partial progress and we commit one chunk at a time. If that's going to happen from each thread then we will need to sync. If we just add the rewritten file set ID to the queue from each thread and synchronize it somewhere else, then this can be removed.

Reply:
Let's prototype how the action is going to do batching commits as discussed yesterday and then adjust this logic accordingly.

Reply:
Removed the locking logic as discussed offline.